sqlite/migrator.go

```diff
@@ -112,6 +119,79 @@ func (m *Migrator) Up(ctx context.Context, source embed.FS) error {
 	return nil
 }
 
+// Down applies the "down" migrations until the SQL database has migrations only >= untilMigration. Use untilMigration = 0 to apply all
+// down migrations, which will delete all data from the database.
+func (m *Migrator) Down(ctx context.Context, untilMigration int, source embed.FS) error {
+	knownMigrations, err := source.ReadDir(".")
+	if err != nil {
+		return err
+	}
+
+	// sort the list according to the version number to ensure the migrations are applied in the correct order
+	sort.Slice(knownMigrations, func(i, j int) bool {
+		return knownMigrations[i].Name() < knownMigrations[j].Name()
+	})
+
+	executedMigrations, err := m.store.allMigrationNames()
+	if err != nil {
+		return err
+	}
+
+	migrationsToDo := len(executedMigrations) - untilMigration
+	if migrationsToDo == 0 {
+		return nil
+	}
+
+	if migrationsToDo < 0 {
+		m.log.Warn("SQL metadata is already on a schema older than target, nothing to do")
+		return nil
+	}
+
+	if m.backupPath != "" {
+		m.log.Info("Backing up pre-migration metadata", zap.String("backup_path", m.backupPath))
+		if err := func() error {
+			out, err := os.Create(m.backupPath)
+			if err != nil {
+				return err
+			}
+			defer out.Close()
+
+			if err := m.store.BackupSqlStore(ctx, out); err != nil {
+				return err
+			}
+			return nil
+		}(); err != nil {
+			return fmt.Errorf("failed to back up pre-migration metadata: %w", err)
+		}
+	}
+
+	m.log.Info("Tearing down metadata migrations", zap.Int("migration_count", migrationsToDo))
+
+	for i := len(executedMigrations) - 1; i >= untilMigration; i-- {
+		downName := knownMigrations[i].Name()
+		downNameNoExtension := dropExtension(downName)
+		upName := executedMigrations[i]
+
+		if downNameNoExtension != upName {
```

Comment:
It's unfortunate that problems here won't get discovered until the assets are already bundled into the binary and `downgrade` is called... instead of checking this here, could we add a new lint script that we run during CI?

Reply:
Added the lint script/check and removed the check here.

If the names don't match, the output will look like this. There's probably room for improvement in the output but hopefully this is clear enough to point to where the mismatch is:

```
------------------------------------------------------------------------------------
Problem detected with SQL migration files: Up and Down migration names do not match!
------------------------------------------------------------------------------------
Diff: Up Migrations without Down Migrations vs. Down Migrations without Up Migrations:
                                                              > sqlite/migrations/0001_create_migrations_table
sqlite/migrations/0006_create_migrations_table                <
```